### PR TITLE
PLANNER-1861 Constraint Verifier documentation

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

**Comment by kie-ci10** [comment]
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

**Comment by triceo** [comment]
ok to test

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 1]
```diff
@@ -574,6 +574,121 @@ the stream.
 ====
 
 
+[[constraintStreamsTesting]]
+== Testing a constraint stream
+
+Constraint streams come with a unit testing harness, called "Constraint Verifier".
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:580]: Don't use ". Use _ for emphasis or * for bold.

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 2]
```diff
@@ -574,6 +574,121 @@ the stream.
...
+[source,java,options="nowrap"]
+----
+    protected Constraint horizontalConflict(ConstraintFactory factory) {
+        return factory
+                .fromUniquePair(Queen.class, equal(Queen::getRowIndex))
+                .penalize("Horizontal conflict", SimpleScore.ONE);
+    }
+----
+
+This is how a simple unit test for this constraint looks, using the Constraint Verifier API:
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:592]: "this is how" = too indirect. Start with the important keywords.
Try something like "A simple unit test for this constraint looks like this:"

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 3]
```diff
@@ -574,6 +574,121 @@ the stream.
...
+        Queen queen2 = new Queen(1, row1, column2);
+        constraintVerifier.verifyThat(NQueensConstraintProvider::horizontalConflict)
+                .given(queen1, queen2)
+                .penalizesBy(1);
+    }
+----
+
+This test ensures that the "Horizontal conflict" constraint gives a penalty of `1` when there are two queens on the
+same row.
+Let's take a look at the test in detail.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:614]: Docs (unlike blogs) should be formal and avoid "Let's".
@emmurphy1 wdyt?

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 4]
```diff
@@ -574,6 +574,121 @@ the stream.
...
+that is being tested, as well as the planning solution (`NQueens`) and the planning entity (`Queen`).
+
+[source,java,options="nowrap"]
+----
+    private ConstraintVerifier<NQueensConstraintProvider, NQueens> constraintVerifier
+            = ConstraintVerifier.build(new NQueensConstraintProvider(), NQueens.class, Queen.class);
+----
+
+The `@Test` annotation indicates that the method is a unit test in a testing framework of your choice.
+Constraint Verifier works with every unit testing framework.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:625]: Every? Are we sure? Also Spock (= very different from junit)?
We could mention Junit, AssertJ and maybe TestNG by name.
How about something like "Constraint Verifier works with JUnit, AssertJ and many others."

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 5]
```diff
@@ -574,6 +574,121 @@ the stream.
...
+
+First part of the test does nothing but prepare the test data.
+In this case, we create two instances of the `Queen` planning entity and their dependencies (`Row`, `Column`):
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:628]: avoid "we"

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 6]
```diff
@@ -574,6 +574,121 @@ the stream.
...
+[source,java,options="nowrap"]
+----
+    constraintVerifier.verifyThat(NQueensConstraintProvider::horizontalConflict)
+            .given(queen1, queen2)
+            .penalizesBy(1);
+----
+
+On the first line, we use the `verifyThat(...)` call to specify a method on the `NQueensConstraintProvider` class which
+we want to test.
+This method must be visible to the test class, which the Java compiler will enforce.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:650]: avoid "will"
use present tense

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc (hunk 1 above)
Comment (emmurphy1) [reply optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:580]: Suggestion:
Constraint streams include the Constraint Verifier unit testing harness.

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc (hunk 2 above)
Comment (emmurphy1) [reply optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:592]: The following example uses the Constraint Verifier API to create a simple unit test for the preceding constrain stream:

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc (hunk 3 above)
Comment (emmurphy1) [reply optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:614]: Yes. Also, don't use " to denote names. I would capitalize Horizontal Conflict. Or you can just make both l/c as I don't see that this is used elsewhere. Either is fine IMO.

This test ensures that the  horizontal conflict constraint assigns a penalty of `1` when there are two queens on the
same row.  

I would just delete the linking text Let's take a look at the test in detail.

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 7]
```diff
@@ -574,6 +574,120 @@ the stream.
...
+        constraintVerifier.verifyThat(NQueensConstraintProvider::horizontalConflict)
+                .given(queen1, queen2)
+                .penalizesBy(1);
+    }
+----
+
+This test ensures that the "Horizontal conflict" constraint gives a penalty of `1` when there are two queens on the
+same row.
+Let's take a look at the test in detail.
+The following line creates a shared `ConstraintVerifier` instance, initializing it with the `NQueensConstraintProvider`
```
Comment (emmurphy1) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:615]: The following line creates a shared `ConstraintVerifier` instance, initializing it with the `NQueensConstraintProvider`. 
What is 'it'? The ContraintVerifier?  

The following line creates a shared `ConstraintVerifier` instance and initialized the instance with the `NQueensConstraintProvider`

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc (hunk 4 above)
Comment (emmurphy1) [reply optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:625]: Mabye:
Constraint Verifier works with many testing freameworks including JUnit and AssertJ. 

In this case 'others' is implied.

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 8]
```diff
@@ -574,6 +574,120 @@ the stream.
...
+[source,java,options="nowrap"]
+----
+    private ConstraintVerifier<NQueensConstraintProvider, NQueens> constraintVerifier
+            = ConstraintVerifier.build(new NQueensConstraintProvider(), NQueens.class, Queen.class);
+----
+
+The `@Test` annotation indicates that the method is a unit test in a testing framework of your choice.
+Constraint Verifier works with every unit testing framework.
+
+First part of the test prepares the test data.
```
Comment (emmurphy1) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:627]: The first part of the test prepares the test data. In this case, the test data includes two instances of the `Queen` planning entity and their dependencies (`Row`, `Column`):

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 9]
```diff
@@ -574,6 +574,120 @@ the stream.
...
+[source,java,options="nowrap"]
+----
+        Row row1 = new Row(0);
+        Column column1 = new Column(0);
+        Column column2 = new Column(1);
+        Queen queen1 = new Queen(0, row1, column1);
+        Queen queen2 = new Queen(1, row1, column2);
+----
+
+Further down, the following code test the constraint itself:
```
Comment (emmurphy1) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:639]: tests the constraint...

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 10]
```diff
@@ -574,6 +574,120 @@ the stream.
...
+    constraintVerifier.verifyThat(NQueensConstraintProvider::horizontalConflict)
+            .given(queen1, queen2)
+            .penalizesBy(1);
+----
+
+The `verifyThat(...)` call is used to specify a method on the `NQueensConstraintProvider` class which is under test.
+This method must be visible to the test class, which the Java compiler will enforce.
+
+The `given(...)` call is used to enumerate all the facts that the constraint stream will operate on.
+In this case, the `queen1` and `queen2` instances previously created.
```
Comment (emmurphy1) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:652]: In this case, the `given(...)` operates on the  `queen1` and `queen2` instances previously created.

Not sure if operates is the correct verb. Substitute or rewrite if need. Just make sure you do include a verb.

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 11]
```diff
@@ -574,6 +574,120 @@ the stream.
...
+Alternatively, you can use a `givenSolution(...)` call here and provide a planning solution instead.
+
+Finally, the `penalizesBy(...)` call completes the test, making sure that the "Horizontal conflict" constraint, given
+one `Queen`, results in a penalty of `1`.
```
Comment (emmurphy1) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:656]: Don't use '' to set apart.

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 12]
```diff
@@ -574,6 +574,120 @@ the stream.
...
+This number is a product of multiplying the match weight, as defined in the constraint stream, by the number of matches.
+
+Alternatively, you can use a `rewardsWith(...)` call to check for rewards instead of penalties.
+The method to use here depends on whether the constraint stream in question is terminated with a `penalize` or a
+`reward` building block.
+
+[[constraintStreamsIntegrationTesting]]
+=== Integration testing
+
+In addition to testing individual constraint, you can also test the entire `ConstraintProvider`.
```
Comment (emmurphy1) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:666]: individual constraints

you can test the entire (remove also, redundant)
...test the entire `ConstraintProvider`(what))?
Noun needed after `ConstraintProvider`

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 13]
```diff
@@ -574,6 +574,120 @@ the stream.
...
+        Queen queen2 = new Queen(1, row2, column2);
+        Queen queen3 = new Queen(2, row3, column3);
+        constraintVerifier.verifyThat()
+                .given(queen1, queen2, queen3)
+                .scores(SimpleScore.of(-3));
+    }
+----
+
+There are only two notable differences to the previous example.
+First, the `verifyThat()` call takes no argument here, signifying that the entire `ConstraintProvider` is being tested.
```
Comment (emmurphy1) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:683]: ConstraintProvider`  (what)

To illustrate, you correctly say the `verifyThat()` call. Need something similar for ConstraintProvider`

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 14]
```diff
@@ -574,6 +574,120 @@ the stream.
...
+Second, instead of either a `penalizesBy()` or `rewardsWith()` call, the `scores(...)` method is used.
+This runs the `ConstraintProvider` on the given facts and returns a sum of `Score`s of all constraint matches resulting
+from the given facts.
+
+Using this method, you can make sure that the constraint provider does not miss any constraints and that the scoring
```
Comment (emmurphy1) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:688]: Minor: I prefer ensure to make sure.

**Comment by triceo** [comment]
Comments resolved and merged, thank you both!

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 15]
```diff
@@ -574,6 +574,122 @@ the stream.
...
+Finally, the `penalizesBy(...)` call completes the test, making sure that the horizontal conflict constraint, given
+one `Queen`, results in a penalty of `1`.
+This number is a product of multiplying the match weight, as defined in the constraint stream, by the number of matches.
+
+Alternatively, you can use a `rewardsWith(...)` call to check for rewards instead of penalties.
+The method to use here depends on whether the constraint stream in question is terminated with a `penalize` or a
+`reward` building block.
+
+[[constraintStreamsIntegrationTesting]]
+=== Integration testing
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc:665]: I am not 100% confortable with calling this "integration testing". It's definitely wider than testing a single constraint unit but it still just testing a ConstraintProvider unit, still far away from end-to-end integration testing on the spectrum.

How about we call this "=== Testing all constraints together" and add a subsection for the first part called "=== Testing each constraint in isolation"?

**Pull request merged**